File: auth_server.py
import os
import json
from flask import Flask, request, jsonify

# Đường dẫn đến tệp lưu thông tin đăng nhập
CREDENTIALS_FILE = os.path.join('translate', 'credentials.json')

# Đường dẫn đến tệp lưu thông tin người dùng
USER_DATA_FILE = os.path.join('translate', 'user_data.json')

# Biến để theo dõi phiên bản dữ liệu
current_data_version = 0

def load_credentials():
    """Hàm tải thông tin đăng nhập từ tệp"""
    if os.path.exists(CREDENTIALS_FILE):
        with open(CREDENTIALS_FILE, 'r', encoding='utf-8') as f:
            return json.load(f)
    return {}

def load_user_data():
    """Hàm tải dữ liệu người dùng từ file"""
    if os.path.exists(USER_DATA_FILE):
        try:
            with open(USER_DATA_FILE, 'r', encoding='utf-8') as f:
                return json.load(f)
        except Exception as e:
            print(f'Lỗi khi đọc file user_data.json: {e}')
            return {
                "usersWindows": [],
                "usersMacOS": [],
                "usersAndroid": [],
                "usersIOS": []
            }
    return {
        "usersWindows": [],
        "usersMacOS": [],
        "usersAndroid": [],
        "usersIOS": []
    }

def save_user_data(data):
    """Hàm lưu dữ liệu người dùng vào file"""
    # Tạo bản sao của dữ liệu để không ảnh hưởng đến dữ liệu gốc
    data_copy = json.loads(json.dumps(data))
    
    # Loại bỏ trường version nếu có
    if 'version' in data_copy:
        del data_copy['version']
    
    # Đảm bảo thư mục tồn tại
    os.makedirs(os.path.dirname(USER_DATA_FILE), exist_ok=True)
    
    with open(USER_DATA_FILE, 'w', encoding='utf-8') as f:
        json.dump(data_copy, f, indent=2, ensure_ascii=False)
    
    print('Dữ liệu đã được lưu vào file user_data.json')
    
    # Đọc lại dữ liệu để xác nhận
    try:
        with open(USER_DATA_FILE, 'r', encoding='utf-8') as f:
            saved_data = json.load(f)
        print(f"Đã lưu {len(saved_data.get('usersWindows', []))} người dùng Windows, "
              f"{len(saved_data.get('usersMacOS', []))} người dùng MacOS, "
              f"{len(saved_data.get('usersAndroid', []))} người dùng Android, "
              f"{len(saved_data.get('usersIOS', []))} người dùng iOS")
    except Exception as e:
        print(f'Lỗi khi đọc lại dữ liệu: {e}')

def find_user_by_account(account):
    """Hàm tìm kiếm người dùng theo tài khoản"""
    if not account:
        return None
    
    # Tải dữ liệu người dùng
    user_data = load_user_data()
    
    # Tìm kiếm trong tất cả các hệ điều hành
    for os_type in ["usersWindows", "usersMacOS", "usersAndroid", "usersIOS"]:
        for user in user_data.get(os_type, []):
            if user.get('account') == account:
                # Đảm bảo người dùng có trường online_status
                if 'online_status' not in user:
                    user['online_status'] = 'Offline'
                
                # Đảm bảo người dùng có trường gps_info
                if 'gps_info' not in user:
                    user['gps_info'] = {
                        'x': '',
                        'y': '',
                        'address': 'Không có'
                    }
                
                # Đảm bảo người dùng có trường wifi_name
                if 'wifi_name' not in user:
                    user['wifi_name'] = 'Không xác định'
                
                return {
                    'user': user,
                    'os_type': os_type
                }
    
    return None

def setup_auth_routes(app):
    """Thiết lập các route cho xác thực và đồng bộ dữ liệu"""
    
    @app.route('/sync_status', methods=['GET'])
    def sync_status():
        try:
            user_data = load_user_data()
            stats = {
                "usersWindows": len(user_data.get("usersWindows", [])),
                "usersMacOS": len(user_data.get("usersMacOS", [])),
                "usersAndroid": len(user_data.get("usersAndroid", [])),
                "usersIOS": len(user_data.get("usersIOS", []))
            }
            return jsonify({
                "status": "success",
                "data_version": current_data_version,
                "stats": stats
            })
        except Exception as e:
            return jsonify({
                "status": "error",
                "message": str(e)
            })
    
    @app.route('/sync_data', methods=['POST'])
    def sync_data():
        global current_data_version
        try:
            data = request.json
            
            # Cập nhật phiên bản dữ liệu
            if 'version' in data:
                current_data_version = data['version']
            
            # Đảm bảo tất cả người dùng có trường gps_info và wifi_name
            for os_type in ["usersWindows", "usersMacOS", "usersAndroid", "usersIOS"]:
                if os_type in data:
                    for user in data[os_type]:
                        if 'gps_info' not in user:
                            user['gps_info'] = {
                                'x': '',
                                'y': '',
                                'address': 'Không có'
                            }
                        if 'wifi_name' not in user:
                            user['wifi_name'] = 'Không xác định'
                        if 'online_status' not in user:
                            user['online_status'] = 'Offline'
            
            # Lưu dữ liệu vào file
            save_user_data(data)
            
            return jsonify({
                "status": "success",
                "message": "Dữ liệu đã được đồng bộ thành công",
                "data_version": current_data_version
            })
        except Exception as e:
            return jsonify({
                "status": "error",
                "message": str(e)
            })
    
    @app.route('/users', methods=['GET'])
    def get_users():
        try:
            user_data = load_user_data()
            return jsonify(user_data)
        except Exception as e:
            return jsonify({
                "status": "error",
                "message": str(e),
                "usersWindows": [],
                "usersMacOS": [],
                "usersAndroid": [],
                "usersIOS": []
            })
    
    @app.route('/update_user_info', methods=['POST'])
    def update_user_info():
        try:
            data = request.json
            account = data.get('account')
            ip_address = data.get('ip')
            gps_info = data.get('gps_info', {})
            wifi_name = data.get('wifi_name')
            online_status = data.get('online_status')
            
            print(f"Nhận yêu cầu cập nhật thông tin cho tài khoản: {account}")
            print(f"Dữ liệu nhận được: {data}")
            
            if not account:
                return jsonify({"status": "error", "message": "Thiếu thông tin tài khoản"})
            
            # Tìm người dùng theo tài khoản
            user_info = find_user_by_account(account)
            
            if not user_info:
                print(f"Không tìm thấy tài khoản {account} trong dữ liệu")
                return jsonify({"status": "error", "message": "Không tìm thấy tài khoản"})
            
            # Tải dữ liệu người dùng
            user_data = load_user_data()
            
            # Lấy thông tin người dùng và hệ điều hành
            user = user_info['user']
            os_type = user_info['os_type']
            
            # Tìm vị trí người dùng trong mảng
            user_index = -1
            for i, u in enumerate(user_data.get(os_type, [])):
                if u.get('account') == account:
                    user_index = i
                    break
            
            if user_index >= 0:
                # Cập nhật thông tin người dùng
                if ip_address:
                    user_data[os_type][user_index]['ip'] = ip_address
                if gps_info:
                    user_data[os_type][user_index]['gps_info'] = gps_info
                if wifi_name:
                    user_data[os_type][user_index]['wifi_name'] = wifi_name
                if online_status:
                    user_data[os_type][user_index]['online_status'] = online_status
                    print(f"Cập nhật trạng thái online thành: {online_status}")
                
                # Lưu dữ liệu người dùng
                save_user_data(user_data)
                print(f"Đã cập nhật thông tin cho tài khoản {account}")
                return jsonify({"status": "success", "message": "Đã cập nhật thông tin người dùng thành công"})
            else:
                print(f"Không tìm thấy tài khoản {account} trong {os_type}")
                return jsonify({"status": "error", "message": "Không tìm thấy tài khoản"})
        except Exception as e:
            print(f"Lỗi khi cập nhật thông tin người dùng: {str(e)}")
            return jsonify({
                "status": "error",
                "message": str(e)
            })
    
    @app.route('/authenticate', methods=['POST'])
    def authenticate():
        try:
            data = request.json
            username = data.get('username')
            password = data.get('password')
            
            print(f"Yêu cầu xác thực cho tài khoản: {username}")
            
            # Kiểm tra thông tin đăng nhập
            credentials = load_credentials()
            
            # Kiểm tra nếu là admin
            if username in credentials and credentials[username] == password:
                print(f"Xác thực admin thành công: {username}")
                return jsonify({
                    "status": "success",
                    "message": "Đăng nhập thành công!",
                    "user": {
                        "username": username,
                        "role": "admin"
                    }
                })
            else:
                # Nếu không phải admin, kiểm tra trong danh sách người dùng
                user_info = find_user_by_account(username)
                
                if user_info:
                    user = user_info['user']
                    os_type = user_info['os_type']
                    
                    if user.get('password') == password:
                        print(f"Xác thực người dùng thành công: {username} ({os_type})")
                        return jsonify({
                            "status": "success",
                            "message": "Đăng nhập thành công!",
                            "user": {
                                "username": username,
                                "account": user.get('account'),
                                "name": user.get('name'),
                                "email": user.get('email', ''),
                                "limited": user.get('limited', 'Unlimited'),
                                "status": user.get('status', 'Active'),
                                "role": "user",
                                "os_type": os_type
                            }
                        })
                
                print(f"Xác thực thất bại cho tài khoản: {username}")
                return jsonify({
                    "status": "error",
                    "message": "Tên đăng nhập hoặc mật khẩu không chính xác!"
                })
        except Exception as e:
            print(f"Lỗi khi xác thực: {str(e)}")
            return jsonify({
                "status": "error",
                "message": f"Lỗi xác thực: {str(e)}"
            }) 

File: translate/auth_server.js
const express = require('express');
const bodyParser = require('body-parser');
const fs = require('fs');
const path = require('path');
const cors = require('cors');
const http = require('http');

const app = express();
const PORT = 5001
;

// Đường dẫn đến tệp lưu thông tin đăng nhập
const CREDENTIALS_FILE = path.join(__dirname, 'credentials.json');

// Đường dẫn đến tệp lưu thông tin người dùng
const USER_DATA_FILE = path.join(__dirname, 'user_data.json');

// Middleware để phân tích cú pháp JSON
app.use(bodyParser.json({ limit: '10mb' }));

// Cấu hình CORS chi tiết
app.use(cors({
    origin: '*', // Cho phép tất cả các nguồn
    methods: ['GET', 'POST', 'PUT', 'DELETE'],
    allowedHeaders: ['Content-Type', 'Authorization']
}));

// Thêm header CORS thủ công cho tất cả các route
app.use((req, res, next) => {
    res.header('Access-Control-Allow-Origin', '*');
    res.header('Access-Control-Allow-Headers', 'Origin, X-Requested-With, Content-Type, Accept, Authorization');
    res.header('Access-Control-Allow-Methods', 'GET, POST, PUT, DELETE, OPTIONS');
    if (req.method === 'OPTIONS') {
        return res.status(200).end();
    }
    next();
});

// Hàm tải thông tin đăng nhập từ tệp
function loadCredentials() {
    if (fs.existsSync(CREDENTIALS_FILE)) {
        const data = fs.readFileSync(CREDENTIALS_FILE, 'utf8');
        return JSON.parse(data);
    }
    return {};
}

// Hàm tải dữ liệu người dùng từ file
function loadUserData() {
    if (fs.existsSync(USER_DATA_FILE)) {
        try {
            const data = fs.readFileSync(USER_DATA_FILE, 'utf8');
            return JSON.parse(data);
        } catch (error) {
            console.error('Lỗi khi đọc file user_data.json:', error);
            return {
                usersWindows: [],
                usersMacOS: [],
                usersAndroid: [],
                usersIOS: []
            };
        }
    }
    return {
        usersWindows: [],
        usersMacOS: [],
        usersAndroid: [],
        usersIOS: []
    };
}

// Thêm biến để theo dõi phiên bản dữ liệu
let currentDataVersion = 0;

// Cập nhật hàm saveUserData để loại bỏ trường version trước khi lưu
function saveUserData(data) {
    // Tạo bản sao của dữ liệu để không ảnh hưởng đến dữ liệu gốc
    const dataCopy = JSON.parse(JSON.stringify(data));
    
    // Loại bỏ trường version nếu có
    delete dataCopy.version;
    
    fs.writeFileSync(USER_DATA_FILE, JSON.stringify(dataCopy, null, 2));
    console.log('Dữ liệu đã được lưu vào file user_data.json');
    
    // Đọc lại dữ liệu để xác nhận
    try {
        const savedData = JSON.parse(fs.readFileSync(USER_DATA_FILE, 'utf8'));
        console.log(`Đã lưu ${savedData.usersWindows.length} người dùng Windows, ${savedData.usersMacOS.length} người dùng MacOS, ${savedData.usersAndroid.length} người dùng Android, ${savedData.usersIOS.length} người dùng iOS`);
    } catch (error) {
        console.error('Lỗi khi đọc lại dữ liệu:', error);
    }
}

// Thêm endpoint kiểm tra trạng thái đồng bộ
app.get('/sync_status', (req, res) => {
    try {
        const userData = loadUserData();
        const stats = {
            usersWindows: userData.usersWindows.length,
            usersMacOS: userData.usersMacOS.length,
            usersAndroid: userData.usersAndroid.length,
            usersIOS: userData.usersIOS.length,
            lastSync: new Date().toISOString(),
            currentVersion: currentDataVersion
        };
        res.status(200).json({
            status: 'success',
            message: 'Trạng thái đồng bộ',
            stats: stats
        });
    } catch (error) {
        res.status(500).json({
            status: 'error',
            message: 'Lỗi khi kiểm tra trạng thái đồng bộ',
            error: error.message
        });
    }
});

// Cập nhật endpoint đồng bộ dữ liệu
app.post('/sync_data', (req, res) => {
    const userData = req.body;
    
    // Kiểm tra dữ liệu hợp lệ
    if (!userData) {
        return res.status(400).json({ status: 'error', message: 'Dữ liệu không hợp lệ' });
    }
    
    try {
        // Đọc dữ liệu hiện tại
        const currentData = loadUserData();
        
        // Kiểm tra phiên bản dữ liệu
        const dataVersion = userData.version || 0;
        
        // Xác định hệ điều hành đang được đồng bộ
        const syncingWindows = userData.hasOwnProperty('usersWindows');
        const syncingMacOS = userData.hasOwnProperty('usersMacOS');
        const syncingAndroid = userData.hasOwnProperty('usersAndroid');
        const syncingIOS = userData.hasOwnProperty('usersIOS');
        
        // Tạo dữ liệu mới bằng cách kết hợp dữ liệu hiện tại và dữ liệu mới
        const newData = {
            usersWindows: syncingWindows ? userData.usersWindows : currentData.usersWindows,
            usersMacOS: syncingMacOS ? userData.usersMacOS : currentData.usersMacOS,
            usersAndroid: syncingAndroid ? userData.usersAndroid : currentData.usersAndroid,
            usersIOS: syncingIOS ? userData.usersIOS : currentData.usersIOS
        };
        
        // Lưu dữ liệu mới
        saveUserData(newData);
        currentDataVersion = dataVersion > currentDataVersion ? dataVersion : currentDataVersion + 1;
        
        // Xác định hệ điều hành đã được đồng bộ
        const syncedOS = [];
        if (syncingWindows) syncedOS.push('Windows');
        if (syncingMacOS) syncedOS.push('MacOS');
        if (syncingAndroid) syncedOS.push('Android');
        if (syncingIOS) syncedOS.push('IOS');
        
        console.log(`Đồng bộ dữ liệu thành công (phiên bản ${currentDataVersion})`);
        console.log(`Đã đồng bộ: ${syncedOS.join(', ')}`);
        
        res.status(200).json({ 
            status: 'success', 
            message: 'Dữ liệu đã được đồng bộ', 
            timestamp: new Date().toISOString(),
            version: currentDataVersion,
            syncedOS: syncedOS,
            userCount: {
                windows: newData.usersWindows.length,
                macos: newData.usersMacOS.length,
                android: newData.usersAndroid.length,
                ios: newData.usersIOS.length
            }
        });
    } catch (error) {
        console.error('Lỗi khi đồng bộ dữ liệu:', error);
        res.status(500).json({ 
            status: 'error', 
            message: 'Lỗi khi đồng bộ dữ liệu', 
            error: error.message 
        });
    }
});

// Endpoint xác thực
app.post('/authenticate', (req, res) => {
    const { username, password } = req.body;
    console.log('Yêu cầu đăng nhập:', username);
    
    // Đọc dữ liệu người dùng
    const userData = loadUserData();
    
    // Kiểm tra trong tất cả các hệ điều hành
    for (const osType of ['usersWindows', 'usersMacOS', 'usersAndroid', 'usersIOS']) {
        const users = userData[osType] || [];
        const user = users.find(u => u.account === username && u.password === password);
        
        if (user) {
            // Kiểm tra trạng thái người dùng
            if (user.status === 'Active') {
                return res.status(200).json({ 
                    status: 'success', 
                    message: 'Đăng nhập thành công!',
                    user: {
                        name: user.name,
                        account: user.account,
                        limited: user.limited,
                        status: user.status,
                        ip: user.ip
                    }
                });
            } else {
                return res.status(401).json({ 
                    status: 'error', 
                    message: `Tài khoản của bạn đang ở trạng thái: ${user.status}` 
                });
            }
        }
    }
    
    // Nếu không tìm thấy người dùng
    res.status(401).json({ status: 'failure', message: 'Tên đăng nhập hoặc mật khẩu không đúng' });
});

// Endpoint lấy dữ liệu người dùng
app.get('/users', (req, res) => {
    const userData = loadUserData();
    res.status(200).json(userData);
});

// Endpoint kiểm tra trạng thái server
app.get('/', (req, res) => {
    res.send('Welcome to the authentication server!');
});

// Endpoint cập nhật địa chỉ IP
app.post('/update_ip', (req, res) => {
    const { account, ip } = req.body;
    const userData = loadUserData();
    
    for (const osType of ['usersWindows', 'usersMacOS', 'usersAndroid', 'usersIOS']) {
        const users = userData[osType] || [];
        const user = users.find(u => u.account === account);
        
        if (user) {
            user.ip = ip;
            saveUserData(userData);
            return res.status(200).json({ status: 'success', message: 'IP updated successfully' });
        }
    }
    
    res.status(404).json({ status: 'error', message: 'User not found' });
});

// Khởi động máy chủ
app.listen(PORT, () => {
    console.log(`Server is running on http://localhost:${PORT}`);
    
    // Tạo file user_data.json nếu chưa tồn tại
    if (!fs.existsSync(USER_DATA_FILE)) {
        const initialData = {
            usersWindows: [],
            usersMacOS: [],
            usersAndroid: [],
            usersIOS: []
        };
        saveUserData(initialData);
        console.log('Đã tạo file user_data.json');
    }
});

// Thêm endpoint cập nhật thông tin người dùng
app.post('/update_user/:userId', (req, res) => {
    const { userId } = req.params;
    const updatedData = req.body;
    
    if (!updatedData) {
        return res.status(400).json({ status: 'error', message: 'Dữ liệu không hợp lệ' });
    }
    
    try {
        const userData = loadUserData();
        let userFound = false;
        
        // Tìm và cập nhật người dùng trong tất cả các hệ điều hành
        for (const osType of ['usersWindows', 'usersMacOS', 'usersAndroid', 'usersIOS']) {
            const users = userData[osType] || [];
            const userIndex = users.findIndex(u => u.name === userId || u.account === userId);
            
            if (userIndex !== -1) {
                // Cập nhật thông tin người dùng
                userData[osType][userIndex] = {
                    ...userData[osType][userIndex],
                    ...updatedData
                };
                userFound = true;
                break;
            }
        }
        
        if (userFound) {
            saveUserData(userData);
            return res.status(200).json({ 
                status: 'success', 
                message: 'Cập nhật thông tin người dùng thành công' 
            });
        } else {
            return res.status(404).json({ 
                status: 'error', 
                message: 'Không tìm thấy người dùng' 
            });
        }
    } catch (error) {
        console.error('Lỗi khi cập nhật thông tin người dùng:', error);
        return res.status(500).json({ 
            status: 'error', 
            message: 'Lỗi khi cập nhật thông tin người dùng', 
            error: error.message 
        });
    }
});
